azure/src/main/java/org/apache/iceberg/azure/adlsv2/VendedAzureSasCredentialProvider.java

```diff
@@ -0,0 +1,197 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class VendedAzureSasCredentialProvider implements Serializable, AutoCloseable {
+  private static final Logger LOG = LoggerFactory.getLogger(VendedAzureSasCredentialProvider.class);
+
+  private final SerializableMap<String, String> properties;
+  private transient volatile Map<String, AzureSasCredentialRefresher>
+      azureSasCredentialRefresherMap;
+  private transient volatile RESTClient client;
+  private transient volatile ScheduledExecutorService refreshExecutor;
+
+  public static final String URI = "credentials.uri";
+  private static final String THREAD_PREFIX = "adls-fileio-credential-refresh";
+
+  public VendedAzureSasCredentialProvider(Map<String, String> properties) {
+    Preconditions.checkArgument(null != properties, "Invalid properties: null");
+    Preconditions.checkArgument(null != properties.get(URI), "Invalid URI: null");
+    this.properties = SerializableMap.copyOf(properties);
+    azureSasCredentialRefresherMap = Maps.newHashMap();
+  }
+
+  public AzureSasCredential getCredential(String storageAccount) {
+    Map<String, AzureSasCredentialRefresher> refresherMap = azureSasCredentialRefresherMap();
+    if (refresherMap.containsKey(storageAccount)) {
+      return refresherMap.get(storageAccount).get();
+    } else {
+      AzureSasCredentialRefresher azureSasCredentialRefresher =
+          new AzureSasCredentialRefresher(
+              () -> this.getSasTokenWithExpiration(storageAccount), credentialRefreshExecutor());
+      refresherMap.put(storageAccount, azureSasCredentialRefresher);
+      return azureSasCredentialRefresher.get();
+    }
+  }
+
+  private Pair<String, Long> getSasTokenWithExpiration(String storageAccount) {
+    LoadCredentialsResponse response = fetchCredentials();
+    List<Credential> adlsCredentials =
+        response.credentials().stream()
+            .filter(c -> c.prefix().contains(storageAccount))
+            .collect(Collectors.toList());
+    Preconditions.checkState(
+        !adlsCredentials.isEmpty(),
+        String.format("Invalid ADLS Credentials for storage-account %s: empty", storageAccount));
+    Preconditions.checkState(
+        adlsCredentials.size() == 1,
+        "Invalid ADLS Credentials: only one ADLS credential should exist per storage-account");
+
+    Credential adlsCredential = adlsCredentials.get(0);
+    checkCredential(adlsCredential, AzureProperties.ADLS_SAS_TOKEN_PREFIX + storageAccount);
+    checkCredential(
+        adlsCredential, AzureProperties.ADLS_SAS_TOKEN_EXPIRE_AT_MS_PREFIX + storageAccount);
+
+    String updatedSasToken =
+        adlsCredential.config().get(AzureProperties.ADLS_SAS_TOKEN_PREFIX + storageAccount);
+    String tokenExpiresAtMillis =
+        adlsCredential
+            .config()
+            .get(AzureProperties.ADLS_SAS_TOKEN_EXPIRE_AT_MS_PREFIX + storageAccount);
+
+    Long expiresAtMs = Long.parseLong(tokenExpiresAtMillis);
```

Comment:
I think you can just inline adlsCredential.config().get(...) inside the Long.parseLong instead of having separate variables.

Reply:
Moved to inline. Initially separated them in an attempt to improve readability but seems redundant as a whole.